release-21.2: [CRDB-9550] kv: adjust number of voters needed calculation when determining replication status

Santamaura opened the pull request:
Backport 2/2 commits from #76430.

/cc @cockroachdb/release

---

Currently, when a range has non-voting replicas and it is queried through replication
stats, it will be reported as underreplicated. This is because in the case where a
zone is configured to have non-voting replicas, for the over/under replicated counts,
we compare the number of current voters to the total number of replicas which is
erroneus. Instead, we will compare current number of voters to the total number of
voters if voters has been set and otherwise will defer to the total number of replicas.
This patch ignores the desired non-voters count for the purposes of this report, for
better or worse. Resolves #69335.

Release justification: low risk bug fix

Release note (bug fix): use total number of voters if set when determining replication
status

Before change:
![Screen Shot 2022-02-11 at 10 03 57 AM](https://user-images.githubusercontent.com/17861665/153615571-85163409-5bac-40f4-9669-20dce77185cf.png)

After change:
![Screen Shot 2022-02-11 at 9 53 04 AM](https://user-images.githubusercontent.com/17861665/153615316-785b156b-bd23-4cfa-a76d-7c9fa47fbf1e.png)

Santamaura committed:
kv: adjust number of voters needed calculation when determining replication status

Before this patch, the replication report's over/under-replication determination
for a range was broken for ranges with non-voting replicas; we were comparing the
desired number of voters against all the current replicas (voters and non-voters).
This patch fixes it by comparing the desired number of voters with the current voters.
In doing so, the over/under-replication counts in the report become defined as relating
exclusively to voters; a range with the wrong number of non-voters is not counted
anywhere in the report. Resolves #69335.

Release note (bug fix): A bug causing incorrect counts of under_replicated_ranges and
over_replicated_ranges in the crdb_internal.replication_stats table for multi-region
databases has been fixed.

Santamaura committed:
kv: change replicas field of zone struct to voters and add nonVoters field
in constraint_stats_report_test.go

Previously, this zone struct had the field replicas which is not granular
enough for the related tests. Instead, the field name is changed to voters
and a new field nonVoters is added.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/84710)
<!-- Reviewable:end -->

Santamaura committed:
kvserver/reports: fix `nil` check in `replicationStatsVisitor`

This fixes a nil pointer dereference in
`replicationStatsVisitor.visitNewZone`.

Release justification: bug fixes and low risk updates to new
functionality.

Release note: None
